contracts/profit/src/state/mod.rs

```diff
@@ -1 +1,231 @@
+            StateEnum::OpenTransferChannel(transfer) => Ok(transfer.config()),
+            StateEnum::OpenIca(_) => Err(ContractError::unsupported_operation(
+                "Querying configuration is not allowed during ICA opening process.",
+            )),
+            StateEnum::Idle(idle) => Ok(idle.config()),
+            StateEnum::BuyBack(_) => Err(ContractError::unsupported_operation(
+                "Querying configuration is not allowed during buy-back.",
+            )),
+        }
+    }
+}
+
+impl From<OpenTransferChannel> for State {
+    fn from(value: OpenTransferChannel) -> Self {
+        Self(StateEnum::OpenTransferChannel(value))
+    }
+}
+
+impl From<IcaConnector> for State {
+    fn from(value: IcaConnector) -> Self {
+        Self(StateEnum::OpenIca(value))
+    }
+}
+
+impl From<Idle> for State {
+    fn from(value: Idle) -> Self {
+        Self(StateEnum::Idle(value))
+    }
+}
+
+impl From<StateLocalOut<BuyBack>> for State {
+    fn from(value: StateLocalOut<BuyBack>) -> Self {
+        Self(StateEnum::BuyBack(value))
+    }
+}
+
+impl ProfitMessageHandler for State {
+    fn confirm_open(
+        self,
+        deps: Deps<'_>,
+        env: Env,
+        channel: Ics20Channel,
+        counterparty_version: String,
+    ) -> ContinueResult<Self> {
+        match self.0 {
+            StateEnum::OpenTransferChannel(transfer) => {
+                transfer.confirm_open(deps, env, channel, counterparty_version)
+            }
+            StateEnum::OpenIca(ica) => ica.confirm_open(deps, env, channel, counterparty_version),
+            StateEnum::Idle(idle) => idle.confirm_open(deps, env, channel, counterparty_version),
+            StateEnum::BuyBack(buy_back) => buy_back
+                .confirm_open(deps, env, channel, counterparty_version)
+                .map(state_machine::from),
+        }
+    }
+}
+
+impl Handler for State {
+    type Response = State;
+    type SwapResult = DexResponse<State>;
+
+    fn on_open_ica(
+        self,
+        counterparty_version: String,
+        deps: Deps<'_>,
+        env: Env,
+    ) -> ContinueResult<Self> {
+        if let StateEnum::OpenIca(ica) = self.0 {
+            ica.on_open_ica(counterparty_version, deps, env)
+        } else {
+            unimplemented!()
+        }
+    }
+
+    fn on_response(self, data: Binary, deps: Deps<'_>, env: Env) -> DexResult<Self> {
+        match self.0 {
+            StateEnum::OpenTransferChannel(transfer) => {
+                transfer.on_response(data, deps, env).map_into()
+            }
+            StateEnum::OpenIca(ica) => ica.on_response(data, deps, env).map_into(),
+            StateEnum::Idle(idle) => idle.on_response(data, deps, env).map_into(),
+            StateEnum::BuyBack(buy_back) => match buy_back.on_response(data, deps, env) {
+                DexResult::Continue(result) => DexResult::Continue(result.map(state_machine::from)),
+                DexResult::Finished(result) => DexResult::Continue(result),
```

Comment:
[few occurences] this is already implemented, Look at packages/dex/src/response.rs for `pub fn map_into<RFrom, HTo, RTo, S>(self) -> Result<HTo>`

Reply:
Actually, it isn't. There it's identity mapping, while here I am mapping both to `Continue` because the types don't match up.